### PR TITLE
Use structs instead of char**, implement GC finalizers, remove deprecated libvlc calls & improve NativePorts callbacks

### PR DESCRIPTION
**Pull request by alexmercerind**
- Separated Dart VM specific headers & libvlcpp into submodules.
- It was necessary to call `Dart_InitializeApiDL` (from Dart using FFI, exported in DLL) for `Dart_NewFinalizableHandle_DL` & there were a lot of linking issues so I decided to include all Dart VM headers/source as opposed to declaring manually within project. This will keep future possible changes from Dart/Flutter team to track easily & updating process easier (maybe?).
- Following C methods have been modified to use structs instead & implemented GC finalizers:
  - [x] `EqualizerCreateEmpty`
  - [x] `EqualizerCreateMode`
  - [x] `MediaParse`
  - [x] `DevicesAll`
- So the project structure is like following:
  - main.cc (dartvlc wrapper)
    - depends on:
      - libvlc (archive)
      - libvlcpp (submodule)
  - api.cc (ffi interface for dartvlc)
    - depends on:
      - dart-api (submodule)
- Removed deprecated `libvlc_video_get_height` & `libvlc_video_get_width` calls.
- Improved NativePort callbacks. No longer deserializing data into strings.

#111

#115

### CODE REVIEW

**Comment by alexmercerind** [comment]
Seems last conflict resolution broke things. Fixing.

File: dartvlc/CMakeLists.txt [hunk 1]
```diff
@@ -8,39 +8,56 @@ endif()
...
+set(DART_VLC_BINARIES "${CMAKE_CURRENT_SOURCE_DIR}/../bin")
+
+set(LIBVLC_ARCHIVE "${DART_VLC_BINARIES}/vlc-${LIBVLC_VERSION}.7z")
+set(LIBVLCPP_ARCHIVE "${DART_VLC_BINARIES}/libvlcpp.zip")
+set(LIBVLC_SOURCE "${EXTERNAL_PACKAGE_DIR}/vlc-${LIBVLC_VERSION}")
+set(LIBVLCPP_SOURCE "${EXTERNAL_PACKAGE_DIR}/libvlcpp-master")
 
-set(LIBVLC_ARCHIVE "${LIBVLC_BINARIES}/vlc-${LIBVLC_VERSION}.7z")
-set(LIBVLCPP_ARCHIVE "${LIBVLC_BINARIES}/libvlcpp.zip")
+set(DARTAPI_ARCHIVE "${DART_VLC_BINARIES}/dart-api.zip")
```
Comment (jnschulze) [inline dartvlc/CMakeLists.txt:21]: I don't see the point of adding this as an archive.
It's just ~5 files as far as I remember. Adding binaries will just unnecessarily complicate change tracking.
Same for libvlcpp (which is quite tiny as well).

If you don't want to add libvlcpp and the dart headers as submodules, let's at least add those files directly as opposed to have CMake unpack them first.

File: dartvlc/api/api.cc [hunk 2]
```diff
@@ -245,123 +234,124 @@ DLLEXPORT void BroadcastCreate(int32_t id, const char* type,
...
-    g_devices_size = 0;
-  }
+void DevicesClear(void*, void* peer) {
+  delete reinterpret_cast<std::vector<const char*>*>(peer);
 }
 
-DLLEXPORT char** EqualizerCreateEmpty() {
+struct DevicesStruct DevicesAll(Dart_Handle object) {
+  auto devices = new std::vector<Device>(Devices::All());
+  auto ids = new std::vector<const char*>();
```
Comment (alexmercerind) [inline dartvlc/api/api.cc:276]: Previously I thought to use `char**` styled C array, but I believe using `std::vector` is a good idea:
- It won't need ask for size inside cleanup finalizer.
- It matches with rest of the code.
- This just requires devices vector to stay in heap along with these two `std::vector<char*>`, which we can easily wipe in finalizer.
- If I settled with `char**`, we'd end up with global variables again (which was our main motive to remove).

File: dartvlc/api/api.cc [hunk 3]
```diff
@@ -200,43 +192,40 @@ DLLEXPORT void PlayerInsert(int32_t id, int32_t index, const char* type,
...
+  Dart_NewFinalizableHandle_DL(
+      object, reinterpret_cast<void*>(metas), sizeof(metas),
+      static_cast<Dart_HandleFinalizer>(MediaClearMap));
+  Dart_NewFinalizableHandle_DL(
+      object, reinterpret_cast<void*>(values), sizeof(values),
+      static_cast<Dart_HandleFinalizer>(MediaClearVector));
+  for (const auto & [ key, value ] : *metas) {
+    values->emplace_back(value.c_str());
   }
+  return values->data();
```
Comment (alexmercerind) [inline dartvlc/api/api.cc:222]: This went cool, got nice `const` at the end :)

File: ffi/lib/src/mediaSource/media.dart [hunk 4]
```diff
@@ -109,9 +109,10 @@ class Media extends MediaSource {
   /// Parses the [Media] to retrieve [Media.metas].
   void parse(Duration timeout) {
     Pointer<Pointer<Utf8>> metas = MediaFFI.parse(
+        this,
         this.mediaType.toString().toNativeUtf8(),
         this.resource.toNativeUtf8(),
-        timeout.inSeconds);
+        timeout.inMilliseconds);
```
Comment (alexmercerind) [inline ffi/lib/src/mediaSource/media.dart:115]: Timeout wasn't working earlier.

File: pubspec.yaml [hunk 5]
```diff
@@ -12,7 +12,8 @@ environment:
 dependencies:
   flutter:
     sdk: flutter
-  dart_vlc_ffi: ^0.1.2
+  dart_vlc_ffi:
+    path: ./ffi
```
Comment (alexmercerind) [inline pubspec.yaml:16]: This is to use current changes. Must be changed back to pub.dev when publishing.

**Review by alexmercerind — COMMENTED** [review]
So, I feel everything is fine & works.
I pushed that Dart VM source code to another personal repo to use as submodule. It has a lot of lines (explanatory comments) & I also checked libvlcpp as submodule now.

**Comment by alexmercerind** [comment]
Few other changes that I'd like to see in future (possibly I can target with this same PR, otherwise feel free to merge):
- Avoiding `CallbackStringArray` for each `NativePort` transfer. I just wanna get rid of deserialization & send data in its "real datatype" instead of `std::to_string` everywhere. Shouldn't be hard as there are only a few events.
- Replacing `libvlc_video_get_height` & `libvlc_video_get_width` with `libvlc_video_get_size` (earlier two are deprecated).

Both the above changes are not blocking. 
I believe we should target macOS build & texture support first.

**Comment by jnschulze** [comment]
Nice, I will try it tomorrow. 


> Few other changes that I'd like to see in future (possibly I can target with this same PR, otherwise feel free to merge):
> 
> * Avoiding `CallbackStringArray` for each `NativePort` transfer. I just wanna get rid of deserialization & send data in its "real datatype" instead of `std::to_string` everywhere. Shouldn't be hard as there are only a few events.
> * Replacing `libvlc_video_get_height` & `libvlc_video_get_width` with `libvlc_video_get_size` (earlier two are deprecated).
>

Yeah, I saw some warnings on macOS. I guess they were suppressed on Windows?

> Both the above changes are not blocking.
> I believe we should target macOS build & texture support first.

Is the macOS build currently broken?

**Comment by alexmercerind** [comment]
> I guess they were suppressed on Windows?
I recall yes. It came to my notice while building on Linux.

> Is the macOS build currently broken?

I don't know if its working (apart from CI result). We should update README from "...Windows & Linux" to "...Windows, Linux & macOS" & close #14 after publishing a new version. You may mention on that issue itself when you feel we're ready as you've actively worked on macOS (and I have no idea about macOS).

**Comment by jnschulze** [comment]
> > Is the macOS build currently broken?
> 
> I don't know if its working (apart from CI result). We should update README from "...Windows & Linux" to "...Windows, Linux & macOS" & close #14 after publishing a new version. You may mention on that issue itself when you feel we're ready as you've actively worked on macOS (and I have no idea about macOS).

I ran the example on macOS and it worked. Videos lagged a bit (on a mid 2013 MacBook Air), but I didn’t have any expectations as we don’t have texture support yet.

**Pull request merged**